diff --git a/kdecore/services/kmimetyperepository.cpp b/kdecore/services/kmimetyperepository.cpp
index f56f48e..65860f7 100644
--- a/kdecore/services/kmimetyperepository.cpp
+++ b/kdecore/services/kmimetyperepository.cpp
@@ -58,7 +58,7 @@ KMimeTypeRepository::~KMimeTypeRepository()
 
 KMimeType::Ptr KMimeTypeRepository::findMimeTypeByName(const QString &_name, KMimeType::FindByNameOption options)
 {
-    QString name = _name;
+    QString name = _name.toLower();
     if (options & KMimeType::ResolveAliases) {
         name = canonicalName(name);
     }
