diff --git a/kded/vfolder_menu.cpp b/kded/vfolder_menu.cpp
index 21c3bbb..450ada6 100644
--- a/kded/vfolder_menu.cpp
+++ b/kded/vfolder_menu.cpp
@@ -1072,7 +1072,7 @@ VFolderMenu::processLegacyDir(const QString &dir, const QString &relDir, const Q
       it.next();
       const QFileInfo fi = it.fileInfo();
       const QString fn = fi.fileName();
-      if (fi.isDir()) {
+      if (fi.isDir() && !fi.isSymLink()) {
          if(fn == QLatin1String(".") || fn == QLatin1String(".."))
             continue;
          SubMenu *parentMenu = m_currentMenu;
@@ -1087,7 +1087,7 @@ VFolderMenu::processLegacyDir(const QString &dir, const QString &relDir, const Q
          m_currentMenu = parentMenu;
          continue;
       }
-      if (fi.isFile() /*&& !fi.isSymLink() ?? */) {
+      if (fi.isFile() && !fi.isSymLink()) {
          if (!fn.endsWith(QLatin1String(".desktop")))
             continue;
          KService::Ptr service;
