--- kdelibs-4.10.5/kio/misc/kpac/CMakeLists.txt.relro	2014-01-08 17:30:13.109957977 +0100
+++ kdelibs-4.10.5/kio/misc/kpac/CMakeLists.txt	2014-01-08 17:30:16.124950458 +0100
@@ -40,6 +40,7 @@ install(TARGETS kded_proxyscout  DESTINA
 set(kpac_dhcp_helper_SRCS kpac_dhcp_helper.c)
 
 kde4_add_executable(kpac_dhcp_helper NOGUI ${kpac_dhcp_helper_SRCS})
+SET_TARGET_PROPERTIES(kpac_dhcp_helper PROPERTIES LINK_FLAGS "-Wl,-z,relro,-z,now")
 
 if (HAVE_NSL_LIBRARY)
   # Assume Solaris
